aerospike/check.py

```diff
@@ -1,77 +1,205 @@
+SINDEX_METRIC_TYPE = '%s.sindex' % SOURCE_TYPE_NAME
+SET_METRIC_TYPE = '%s.set' % SOURCE_TYPE_NAME
+
+Addr = namedtuple('Addr', ['host', 'port'])
+
+def parse_namespace(data, namespace, secondary):
+    idxs = []
+    while data != []:
+        l = data.pop(0)
+
+        # $ asinfo -v 'sindex/phobos_sindex'
+        # ns=phobos_sindex:set=longevity:indexname=str_100_idx:num_bins=1:bins=str_100_bin:type=TEXT:sync_state=synced:state=RW; \
+        # ns=phobos_sindex:set=longevity:indexname=str_uniq_idx:num_bins=1:bins=str_uniq_bin:type=TEXT:sync_state=synced:state=RW; \
+        # ns=phobos_sindex:set=longevity:indexname=int_uniq_idx:num_bins=1:bins=int_uniq_bin:type=INT SIGNED:sync_state=synced:state=RW;
 
+        # $ asinfo -v 'sets/bar'
+        # ns=bar:set=demo:objects=1:tombstones=0:memory_data_bytes=34:truncate_lut=0:stop-writes-count=0:set-enable-xdr=use-default:disable-eviction=false;\
+        # ns=bar:set=demo2:objects=123456:tombstones=0:memory_data_bytes=8518464:truncate_lut=0:stop-writes-count=0:set-enable-xdr=use-default:disable-eviction=false
+
+        match = re.match('^ns=%s:([^:]+:)?%s=([^:]+):.*$' % (namespace, secondary), l)
+        if match == None:
+            continue
+        idxs.append(match.groups()[1])
+
+    return idxs   
 
 class AerospikeCheck(AgentCheck):
 
     def __init__(self, name, init_config, agentConfig, instances=None):
         AgentCheck.__init__(self, name, init_config, agentConfig, instances)
-        self.connection_pool = {}
+        self.connections = {}
 
     def check(self, instance):
-        """Run the Aerospike check for one instance."""
+        addr, metrics, namespace_metrics, required_namespaces, tags = \
+            self._get_config(instance)
 
-        # Required parameters
-        host = instance['host']
-        port = int(instance['port'])
+        try:
+            conn = self._get_connection(addr)
 
-        # Optional parameters
-        want_metrics = set(instance.get('metrics', []))
-        want_namespace_metrics = set(instance.get('namespace_metrics', []))
-        namespaces = instance.get('namespaces', None)
+            with closing(conn.makefile('r')) as fp:
+                conn.send('statistics\r')
+                self._process_data(fp, CLUSTER_METRIC_TYPE, metrics, tags=tags)
+
+                namespaces = self._get_namespaces(conn, fp, required_namespaces)
+
+                for ns in namespaces:
+                    conn.send('namespace/%s\r' % ns)
+                    self._process_data(fp, NAMESPACE_METRIC_TYPE, namespace_metrics, tags+['namespace:%s' % ns])
+
+                    conn.send('sindex/%s\r' % ns)
+                    for idx in parse_namespace(fp.readline().split(';')[:-1], ns,'indexname'):
+                        conn.send('sindex/%s/%s\r' % (ns,idx))
+                        self._process_data(fp, SINDEX_METRIC_TYPE, [], 
+                                tags+['namespace:%s' % ns, 'sindex:%s.%s' % (ns,idx)])
```

Comment:
if we're already sending `namespace:foo` does it make sense to use `foo` in `sindex` (ie. `sindex:foo.1`)? Or would sending `namespace:foo` and `sindex:1` suffice? Also, again to understand cardinality, what would be the expected order of magnitude for `sindex`?

Reply:
Max of 256 sindexs per namespace. 

We can have the same sindex name across different namespaces.

You're asking about the tagging, right? Why have `namespace:foo` and `sindex:foo.bar` instead of `namespace:foo.bar`? I'm not entirely sure of DD's front end, but generally having 2 separate tags would mean I can have a partial match for metrics, instead of exact match.

Reply:
I actually mean why have `namespace:foo` `sindex:foo.bar` instead of `namespace:foo`, `sindex:bar`?

Reply:
Ah ok.